src/mcfret.c

```diff
@@ -0,0 +1,621 @@
+    for (samples=non->begin;samples<non->end;samples++){ 
+        ti=samples*non->sample;
+        if (non->cluster!=-1){
+            if (read_cluster(non,ti,&cl,Cfile)!=1){
+	              printf("Cluster trajectory file to short, could not fill buffer!!!\n");
+	              printf("ITIME %d\n",ti);
+	              exit(1);
+            }
+            /* Configuration belong to cluster */ 
+            if (non->cluster==cl){
+	              Ncl++;
+            }
+        }
+        if (non->cluster==-1 || non->cluster==cl){
+            /* Read Hamiltonian */
+            read_Hamiltonian(non,Hamil_i_e,H_traj,ti);
+          	    
+            /* Extract couplings between segments */
+            multi_projection_Coupling(Hamil_i_e,non);
+            for (i=0;i<non->singles;i++){
+              for (j=i+1;j<non->singles;j++){
+                J[non->singles*i+j]+=Hamil_i_e[Sindex(i,j,non->singles)];
+                J[non->singles*j+i]+=Hamil_i_e[Sindex(i,j,non->singles)];
+              }
+            }
+        } /* We are closing the cluster loop */
+
+         /* Update NISE log file */ 
+        log=fopen("NISE.log","a");
+        fprintf(log,"Finished sample %d\n",samples);
+          
+        time_now=log_time(time_now,log);
+        fclose(log);
+    }/* Closing the loop over samples */
+    
+    /* Divide with total number of samples */
+    for (i=0;i<non->singles;i++){
+        for (j=0;j<non->singles;j++){
+            J[non->singles*i+j]=J[non->singles*i+j]/samples;
+        }
+    }
+
+    /* The calculation is finished, lets write output */
+    log=fopen("NISE.log","a");
+    fprintf(log,"Finished Calculating Response!\n");
+    fprintf(log,"Writing to file!\n");  
+    fclose(log);
+
+    if (non->cluster!=-1){
+        printf("Of %d samples %d belonged to cluster %d.\n",samples,Ncl,non->cluster);
+        if (samples==0){ /* Avoid dividing by zero */ 
+            samples=1;
+        }
+    }
+
+    fclose(mu_traj),fclose(H_traj);
+    if (non->cluster!=-1){
+        fclose(Cfile);
+    }
+    free(Hamil_i_e);
+    free(mu_xyz);  
+    return;
+
+}
+
+
+/* Find MCFRET segments using an automatic scheme */
+void mcfret_autodetect(t_non *non, float treshold);
+
+/* Calculate actual rate matrix */
+void mcfret_rate(float *rate_matrix,float *coherence_matrix,int segments,float *re_Abs,float *im_Abs,
+    float *re_Emi,float *im_Emi,float *J,t_non *non){
+    int nn2,N;
+    int si,sj;
+    int i,j,k;
+    int *ns; /* Segment dimensions */
+    int t1;
+    float *rate_response;
+    float rate;
+    float isimple,is13; /* Variables for integrals */
+    float *re_aux_mat,*im_aux_mat;
+    float *re_aux_mat2,*im_aux_mat2;
+    float *Zeros;
+    FILE *ratefile;
+    N=non->singles;
+    nn2=non->singles*non->singles;
+
+    rate_response=(float *)calloc(non->tmax,sizeof(float));
+    re_aux_mat=(float *)calloc(nn2,sizeof(float));
+    im_aux_mat=(float *)calloc(nn2,sizeof(float));
```

Comment:
Many of theses variables are not freed at the end

Reply:
Good point! Implementing!